THREESCALE-11526 - Redis Sentinel with TLS for Porta and Backend (#1041)
3scale/3scale-operator: doc/redis-sentinel-tls.md

      - [Create Redis and Sentinel secrets](#create-redis-and-sentinel-secrets)
    - [Install Redis Server with TLS support](#install-redis-server-with-tls-support)
      - [Redis ConfigMap:](#redis-configmap)
      - [Redis StatefulSet](#redis-statefulset)
    - [Install Sentinel with TLS support](#install-sentinel-with-tls-support)
      - [Sentinel Service](#sentinel-service)
      - [Sentinel ConfigMap](#sentinel-configmap)
      - [Sentinel StatefulSet](#sentinel-statefulset)
    - [Installation Results](#installation-results)
    - [Summary](#summary)
  - [The Process of Client interaction with the Sentinel and Redis server](#the-process-of-client-interaction-with-the-sentinel-and-redis-server)
    - [Role of Sentinel](#role-of-sentinel)
    - [How Clients Interact with Sentinel and Redis Servers](#how-clients-interact-with-sentinel-and-redis-servers)
    - [How Failover Works with Sentinel](#how-failover-works-with-sentinel)
    - [Summary](#summary-1)
  - [Sentinel Test on EC2 (briefly)](#sentinel-test-on-ec2-briefly)
  - [Some useful commands](#some-useful-commands)

## Overview

### Redis and Redis-Sentinel (Sentinel): Same Image, Different Configuration

> *We will use term **Sentinel** instead of full name **Sentinel** in most places in this doc.*

Reviewer: Redis Sentinel perhaps?

Author: Thank you. Done.

#### Key Points
- **Same image**: Redis and Sentinel share the same Redis image.
- **Different configurations**: Redis runs as a normal database, while Sentinel is configured and run as a process to monitor and manage Redis instances.
- **Redis and Sentinel are typically run as separate processes**: You run one or more Redis instances and multiple Sentinel instances (usually 3 or more) for high availability.

 **Sentinel** is responsible for monitoring the Redis instance and handling failover in case the primary Redis instance goes down. You can scale the number of Sentinel nodes and configure your application to connect to Redis using the Sentinel service for high availability.

If both the **Redis server** (redis) and **Sentinel** (redis-sentinel) are using **TLS** for communication, you'll need to configure both the Redis server and the Sentinel to use **SSL/TLS encryption** for secure connections. This involves setting up certificates and keys for both the Redis server and the Sentinel, and then ensuring the communication between them and the clients is secured via TLS.

Here’s how to do this step-by-step in the OpenShift environment.

> The Guide includes not only Sentinel TLS configuration, but Redis server also, to be useful for Testing. 

Reviewer: ```suggestion
The guide includes both Sentinel TLS configuration and Redis server, which are useful for testing. 
```

Author: Thank you for suggestion. Yes, need change it. I  actually wanted to say before - that the Guide is for Sentinel. But there is No Sentinel without Redis server, so providing Redis-server configuration also. So after "consulting" with ChatGPT  I changed it to "The guide includes both Sentinel TLS configuration and Redis server setup, which is necessary for testing Sentinel. "   Hope it looks better now.

It will also include failover scenarion verification.

##  Sentinel environment setup

### Prepare TLS Certificates
Before you can configure Redis and Sentinel to use TLS, you'll need to create and configure certificates for the Redis server, Sentinel, and optionally for the client.

> **Note**: To correctly generate the Redis server and Sentinel certificates, you need to use their services IPs as the Common Names (CN) in the certificates. If you haven’t set up the services yet, please refer to **Section 4** to create the Redis server and Sentinel services, before continue to next item. Once the services are running, use their respective IPs as the CN when generating the certificates.

Author: Updated. Thank you

redis   ClusterIP   172.30.115.228   <none>        6379/TCP,6380/TCP   6s
```

#### Generate SSL Certificates (Self-signed or CA-signed)
You can create self-signed certificates or use certificates signed by a Certificate Authority (CA). Below is an example of generating self-signed certificates:
- Generate the CA private key and certificate
- Generate Redis server private key, CSR (Certificate Signing Request) and certificate
- Generate Sentinel private key, CSR and certificate
- Generate Redis Client private key, CSR and certificate

Folloing example contains services IPs, please change it to yours.

```bash
openssl genpkey -algorithm RSA -out ca.key -pkeyopt rsa_keygen_bits:2048
openssl req -x509 -new -key ca.key -out ca.crt -days 3650 -subj "/CN=Redis CA"

openssl genpkey -algorithm RSA -out redis-server.key -pkeyopt rsa_keygen_bits:2048
openssl req -new -key redis-server.key -out redis-server.csr -subj "/CN=172.30.115.228"
openssl x509 -req -in redis-server.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out redis-server.crt -days 365

openssl genpkey -algorithm RSA -out redis-sentinel.key -pkeyopt rsa_keygen_bits:2048

> openssl req -new -key redis-sentinel.key -out redis-sentinel.csr -subj "/CN=redis-sentinel.3scale-test.svc.cluster.local"

Reviewer: ^^ that comment makes no sense lol.
Basically, whenever we are mentioning DNS, maybe it would be better to expose DNS as part of your setup to run these commands and just use that instead? It's easy to miss it when doing for the first time

> openssl req -new -key redis-sentinel.key -out redis-sentinel.csr -subj "/CN=redis-sentinel.3scale-test.svc.cluster.local"

Author: yes, meanwhile DNS is not working for master monitorying by Sentinel, so using IP.  "issues found" section added (see before). Thank you.

openssl x509 -req -in redis-sentinel.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out redis-sentinel.crt -days 365

openssl genpkey -algorithm RSA -out redis-client.key
openssl req -new -key redis-client.key -out redis-client.csr -subj "/CN=redis-client.example.com"
openssl x509 -req -in redis-client.csr -CA ca.crt -CAkey ca.key -CAcreateserial -out redis-client.crt -days 365
```

This will generate the following certificates and keys:
- `ca.crt`: The CA certificate
- `redis-server.crt` and `redis-server.key`: The Redis server’s certificate and key
- `redis-sentinel.crt` and `redis-sentinel.key`: The Sentinel’s certificate and key

> - `redis-client.crt` and `redis-client.key`: The Redis client's certificate and key

Author: Updated/added:
+- Other (intermediate) files, such as *.csr and ca.srl, are also generated. If needed, additional information can be found in the OpenSSL documentation.
Thank you

```bash
oc create secret generic redis-tls-secret \
  --from-file=redis-server.crt \
  --from-file=redis-server.key \
  --from-file=ca.crt

oc create secret generic redis-sentinel-tls-secret \
  --from-file=redis-sentinel.crt \
  --from-file=redis-sentinel.key \
  --from-file=ca.crt
```

These secrets will contain the necessary certificates for Redis and Sentinel.

### Install Redis Server with TLS support

#### Redis ConfigMap:

```yaml
cat << EOF | oc create -f -
apiVersion: v1
kind: ConfigMap
metadata:
  name: redis-config-redis  
data:
  redis.conf: |+
    # redis.conf
    bind 0.0.0.0
    protected-mode no
    port 6379
    tls-port 6380
    tls-cert-file /etc/redis/certs/redis-server.crt
    tls-key-file /etc/redis/certs/redis-server.key
    tls-ca-cert-file /etc/redis/certs/ca.crt
    tls-auth-clients no
    stop-writes-on-bgsave-error no
    save ""
EOF
```

#### Redis StatefulSet

```yaml
cat << EOF | oc create -f -
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: redis
spec:
  serviceName: "redis"
  replicas: 3
  selector:
    matchLabels:
      app: redis
  template:
    metadata:
      labels:
        app: redis
    spec:
      containers:
      - name: redis

>         #image: quay.io/fedora/redis-6

Reviewer: can be removed?

Author: Done. Thank you

>         image: redis:7

Reviewer: lets move out of dockerhub - mirror.gcr.io/library/redis:7

            mountPath: /etc/redis/certs
            readOnly: true
        command: ["/bin/sh", "-c", "redis-server /etc/redis/redis.conf"]
      volumes:
        - name: redis-config-volume
          configMap:
            name: redis-config-redis
        - name: redis-tls-volume
          secret:
            secretName: redis-tls-secret
EOF
```

### Install Sentinel with TLS support

#### Sentinel Service

```yaml
cat << EOF | oc create -f -
apiVersion: v1
kind: Service
metadata:
  name: redis-sentinel
spec:
  selector:
    app: redis-sentinel
  ports:
    - name: redis 
      protocol: TCP
      port: 26379 
      targetPort: 26379 
    - name: redis-tls
      protocol: TCP
      port: 26380 
      targetPort: 26380
  clusterIP: None  # This makes the service headless
EOF
```

#### Sentinel ConfigMap
  
```yaml
cat << EOF | oc create -f -
apiVersion: v1
kind: ConfigMap
metadata:
  name: redis-sentinel-config 
data:
  sentinel.conf: |
    port 0
    tls-port 26380
    tls-cert-file /etc/redis/certs/redis-sentinel.crt
    tls-key-file /etc/redis/certs/redis-sentinel.key
    tls-ca-cert-file /etc/redis/certs/ca.crt
    tls-auth-clients yes 
    tls-replication yes

>     #sentinel monitor mymaster redis.3scale-test.svc.cluster.local 6380 2

Reviewer: might be worth to explain that this is using the DNS instead and is recommended. My ns is different so maybe something like "sentinel monitor mymaster redis.<NAMESPACE>.svc.cluster.local 6380 2"

Reviewer: I've tried it with dns and it failed.

Author: I tried more but didn't succed to work with DNS for master monitoring . 
New section added : 
![image](https://github.com/user-attachments/assets/dfe6fa8b-a0b8-49ed-9b3a-7cb38013e68b)

- If you will define both `port 26379` (default) and `tls-port 26380`   - Sentinel will use both TLS and non TLS communications. We found that in logs it looks like Sentinel has two replicas that connect to Redis, TLS and non-Tls (details are not provided here).

#### Sentinel StatefulSet

```yaml
cat << EOF | oc create -f -
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: redis-sentinel
spec:
  serviceName: "redis-sentinel"
  replicas: 3
  selector:
    matchLabels:
      app: redis-sentinel
  template:
    metadata:
      labels:
        app: redis-sentinel
    spec:
      initContainers:
        - name: copy-sentinel-config
          image: quay.io/fedora/redis-6
          command: ["/bin/sh", "-c", "cp /etc/redis/sentinel.conf /mnt/config/sentinel.conf"]
          volumeMounts:
            - name: redis-config-volume
              mountPath: /etc/redis/sentinel.conf
              subPath: sentinel.conf  # Read-only mount from ConfigMap
            - name: sentinel-config-writable
              mountPath: /mnt/config  # Writable volume for copying the file
      containers:
        - name: redis-sentinel
          #image: quay.io/fedora/redis-6

>           image: redis:7

Reviewer: mirror.gcr.io/library/redis:7 and I think I would remove the above comment

Author: done. Thank you

          ports:

>             - containerPort: 26379

Reviewer: In above cm we only have tls-port 26380 - do we need 26379 as well here?

Author: non-TLS port can be useful for debugging in Testing environment. 
Thank you

So, **client applications never directly communicate with the Sentinel** for data operations. 
They always connect to the Redis server (master) that Sentinel has informed them about. 
The role of Sentinel is to monitor Redis instances, handle failover, and provide the current master’s address to the client.  
Please see [operator-user-guide.md](./operator-user-guide.md) for SENTINEL configuration in 3scale.  

- We found it challenging to observe Sentinel's failover process in Kubernetes, as K8S/OpenShift resolves the issue faster than Sentinel. Therefore, we tested Sentinel's behavior on EC2, as detailed in the next section.

## Sentinel Test on EC2 (briefly)
The goal of this test on EC2 was to observe the failover process with Sentinel. As mentioned earlier, it's challenging to 
observe Sentinel's failover process in Kubernetes because K8S/OpenShift handles the issue faster than Sentinel 
can decide which replica to promote as the Redis master.  
We use non-TLS configuration for simplicity.  

Following was done:
- EC2 Ubuntu instance was created
- Redis sources were downloaded, compiled, installed, run.
- Sentinel's failover handling of redis was observed.  

> - We are providing only logs here. For installation details of Redis+Sentinel on EC2 - please see AWS and Redis docs or ChatGPT.

Reviewer: ```suggestion
- We are providing only logs here. For installation details of Redis+Sentinel on EC2 - please see AWS and Redis docs.
```

Author: updated. Thank you
